adapters/saml/core/src/main/java/org/keycloak/adapters/saml/PropertiesBasedRoleMapper.java

```diff
@@ -0,0 +1,177 @@
+ * it maps to a set of one ore more different roles, then these roles are set in the result set. If no mapping is found for the role
+ * then it is included as is in the result set.
+ *
+ * Once the roles have been processed, the implementation checks if the principal extracted from the assertion contains an entry
+ * in the {@code properties} file. If a mapping for the principal exists, any roles listed as value are added to the result set. This
+ * allows the assignment of extra roles to a principal.
+ *
+ * For example, consider the following {@code role-mappings.properties}:
+ *
+ * <pre>
+ *     # role to roles mappings
+ *     samlRoleA=jeeRoleX,jeeRoleY
+ *     samlRoleB=
+ *
+ *     # principal to roles mappings
+ *     kc-user=jeeRoleZ
+ * </pre>
+ *
+ * If the {@code {@link #map(String, Set)}} method is called with {@code kc-user} as principal and a set containing roles
+ * {@code samlRoleA,samlRoleB,samlRoleC}, the result set will be formed by the roles {@code jeeRoleX,jeeRoleY,samlRoleC,jeeRoleZ}.
+ * In this case, {@code samlRoleA} is mapped to two roles ({@code jeeRoleX,jeeRoleY}), {@code samlRoleB} is discarded as it is
+ * mapped to an empty role, {@code samlRoleC} is used as is and the principal is also assigned {@code jeeRoleZ}.
+ *
+ * @author <a href="mailto:sguilhen@redhat.com">Stefan Guilhen</a>
+ */
+public class PropertiesBasedRoleMapper implements RoleMappingsProvider {
+
+    private static final Logger logger = Logger.getLogger(PropertiesBasedRoleMapper.class);
+
+    public static final String PROVIDER_ID = "properties-based-role-mapper";
+
+    private static final String PROPERTIES_LOCATION = "properties.file.location";
+
+    private static final String DEFAULT_LOCATION = "/WEB-INF/role-mappings.properties";
+
+    private Properties roleMappings;
+
+    @Override
+    public String getId() {
+        return PROVIDER_ID;
+    }
+
+    @Override
+    public void init(final SamlDeployment deployment, final ResourceLoader loader, final Properties config) {
+        String path = DEFAULT_LOCATION;
+        if (config.getProperty(PROPERTIES_LOCATION) != null) {
+            path = config.getProperty(PROPERTIES_LOCATION);
+        }
+        InputStream is = loader.getResourceAsStream(path);
+        // load properties file - first try to find the file in the WAR using the resoure loader, if that fails look in the filesystem.
+        this.roleMappings = new Properties();
+        boolean loaded = false;
+        try {
+            this.roleMappings.load(is);
+            loaded = true;
+            if (logger.isDebugEnabled()) {
+                logger.debugf("Resource loader successfully loaded role mappings from %s", path);
+            }
+        } catch(Exception e) {
+            if (logger.isDebugEnabled()) {
+                logger.debugv(e, "Resource loader unable to load role mappings from %s", path);
+            }
+        }
+        if (!loaded && !DEFAULT_LOCATION.equals(path)) {
```

Comment:
This is something that crossed my mind too when I was coding this part. You have a good point there.

Thinking about this now and looking into the xsd I found something interesting: the key-store-type locates the keystore using two separate attributes (one for filesystem, one for a WAR resource). It might be a good idea to use two different config options here as well - something like filesystem_location and resource_location - and use both when trying to locate the properties file. We can still have a default one pointing to the WAR resource WEB-INF but that would not interfere with the filesystem config if there is one. WDYT?

Reply:
Hmm, I am probably ok with both. If you decide that what you suggested above is not worth doing (it will require changes in docs, code, tests) I will not insist on changing it. Personally, I like having two config options a little bit more.